FirebaseAppCheck/Sources/AppAttestProvider/FIRAppAttestProvider.m

```diff
@@ -118,18 +126,51 @@ - (nullable instancetype)initWithApp:(FIRApp *)app {
 #pragma mark - FIRAppCheckProvider
 
 - (void)getTokenWithCompletion:(void (^)(FIRAppCheckToken *_Nullable, NSError *_Nullable))handler {
+  [self getToken]
+      // Call the handler with the result.
+      .then(^FBLPromise *(FIRAppCheckToken *token) {
+        handler(token, nil);
+        return nil;
+      })
+      .catch(^(NSError *error) {
+        handler(nil, error);
+      });
+}
+
+- (FBLPromise<FIRAppCheckToken *> *)getToken {
+  // Check attestation state to decide on the next steps.
+  return [self attestationState].thenOn(self.queue, ^id(FIRAppAttestProviderState *attestState) {
+    switch (attestState.state) {
+      case FIRAppAttestAttestationStateUnsupported:
+        return attestState.appAttestUnsupportedError;
+        break;
+
+      case FIRAppAttestAttestationStateSupportedInitial:
+      case FIRAppAttestAttestationStateKeyGenerated:
+        // Initial handshake is required.
+        return [self initialHandshakeWithKeyID:attestState.appAttestKeyID];
+        break;
+
+      case FIRAppAttestAttestationStateKeyRegistered:
+
```

Comment:
optional: maybe add a comment like the one on line 150, e.g. `// Token refresh is required.` (I noticed the newline here so purely suggesting!)